tests/ui/pattern/usefulness/impl-trait.rs

```diff
@@ -0,0 +1,135 @@
+#![feature(never_type)]
+#![feature(exhaustive_patterns)]
+#![feature(type_alias_impl_trait)]
+#![feature(non_exhaustive_omitted_patterns_lint)]
+#![deny(unreachable_patterns)]
+// Test that the lint traversal handles opaques correctly
+#![deny(non_exhaustive_omitted_patterns)]
+
+fn main() {}
+
+#[derive(Copy, Clone)]
+enum Void {}
+
+fn return_never_rpit(x: Void) -> impl Copy {
+    if false {
+        match return_never_rpit(x) {}
+    }
+    x
+}
+fn friend_of_return_never_rpit(x: Void) {
+    match return_never_rpit(x) {}
+}
```

Comment:
Hum, something weird happened after I rebased on a more recent master: before, `typeck_results.concrete_opaque_types` only had a revealed type for an opaque if we were in a defining use. This seems to no longer be the case. `friend_of_return_never_rpit` above used to error "impl Copy is not empty" as it should, and after the rebase it doesn't :'(

@compiler-errors do you happen to know how I can check if the thir body I am inspecting is within a defining use of a given opaque?

Reply:
I checked out your PR and this function emits:

```
error[E0004]: non-exhaustive patterns: type `impl Copy` is non-empty
  --> /home/gh-compiler-errors/test.rs:21:11
   |
21 |     match return_never_rpit(x) {}
   |           ^^^^^^^^^^^^^^^^^^^^
   |
   = note: the matched value is of type `impl Copy`
help: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern as shown
   |
21 ~     match return_never_rpit(x) {
22 +         _ => todo!(),
23 +     }
```

Reply:
This may be due to #117416. Try reverting it and seeing if it restores your error message?

Reply:
Ah, so as you suggested the issue comes from other errors making the compiler not continue with checking things. No opaque type was revealed more than it should; `typeck_results.concrete_opaque_types` is still reliable.

Reverting your PR did not restore my error message. However, removing [this part](https://github.com/Nadrieril/rust/blob/7dca8df557bbede09ee24c1d3e491e03edf21378/tests/ui/pattern/usefulness/impl-trait.rs#L47-L50) of an unrelated function body did fix it! So all is good.

No longer relevant to this PR, but: I'm pretty surprised by the action-at-a-distance here, any idea what caused it?

Reply:
Actually if you had a clue that could simplify testing. If you don't I'll just make a bunch of separate test files

Reply:
Tests are fixed